twister: allow list input for set config fields

gmarull opened the pull request:
Twister configuration files have a few fields defined with the "set"
type. The set needs to be specified as a space-separated list of
entries, for example:

```yaml
plaform_allow: boardx boardy
```

This patch allows users to write these fields as YAML lists, for
example:

```yaml
platform_allow:
  - boardx
  - boardy
```

The list gets converted to a Python set, therfore the same exact
behavior is kept in Twister.

In some cases like in the example above, using lists allows to have more
readable configuration files, specially if the list is long. It also
helps on diffs, provided we have one item per line.

Note: pykyalify doesn't support "multi-type" fields, so we either switch
to the "any" type, or, we move to JSON schema. See efefbc8d2315a2dd97527a91e661bbc1766eca3d

gmarull committed:
twister: allow list input for set config fields

Twister configuration files have a few fields defined with the "set"
type. The set needs to be specified as a space-separated list of
entries, for example:

```yaml
plaform_allow: boardx boardy
```

This patch allows users to write these fields as YAML lists, for
example:

```yaml
platform_allow:
  - boardx
  - boardy
```

The list gets converted to a Python set, therfore the same exact
behavior is kept in Twister.

In some cases like in the example above, using lists allows to have more
readable configuration files, specially if the list is long. It also
helps on diffs, provided we have one item per line.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

nashif commented:
> In some cases like in the example above, using lists allows to have more readable configuration files, specially if the list is long. It also helps on diffs, provided we have one item per line.
> 

agree.

> Note: pykyalify doesn't support "multi-type" fields, so we either switch to the "any" type, or, we move to JSON schema. See [efefbc8](https://github.com/zephyrproject-rtos/zephyr/commit/efefbc8d2315a2dd97527a91e661bbc1766eca3d)

or maybe convert all existing yaml files to use the new format, and while we are at it, convert other keys with same behaviour, because platform_allow is not the only one...
but as usual, this will break out of tree tests, so we need a way round this, main reason it was not done until now..

Also, we were considering renaming some of those entries in the tests, i.e. platform_allow, which has been confusing for many users/developers

@PerMac  your thoughts.

gmarull commented:
> > In some cases like in the example above, using lists allows to have more readable configuration files, specially if the list is long. It also helps on diffs, provided we have one item per line.
> 
> agree.
> 
> > Note: pykyalify doesn't support "multi-type" fields, so we either switch to the "any" type, or, we move to JSON schema. See [efefbc8](https://github.com/zephyrproject-rtos/zephyr/commit/efefbc8d2315a2dd97527a91e661bbc1766eca3d)
> 
> or maybe convert all existing yaml files to use the new format, and while we are at it, convert other keys with same behaviour, because platform_allow is not the only one... but as usual, this will break out of tree tests, so we need a way round this, main reason it was not done until now..
> 
> Also, we were considering renaming some of those entries in the tests, i.e. platform_allow, which has been confusing for many users/developers
> 
> @PerMac your thoughts.

To avoid breaking behavior, I'd support both types, str and list (at least for a while). But this implies using the `any` type, or, move to JSON schema which allows multiple types AFAIK.

gmarull commented:
Btw, we could also consider adding new replacement fields and deprecate the old ones, e.g. `allow_platforms` -> `list`, `platform_allow` -> `set` (deprecated).

hakehuang commented on tests/subsys/pm/policy_api/testcase.yaml:
```diff
@@ -3,7 +3,9 @@
 
 common:
   tags: pm
-  platform_allow: native_posix native_posix_64
```
as there are still many legacy testcases in old format, maybe it is better to add a new case, and keep old ones, as I assume the two format will co-exist for some time.

nashif commented:
> To avoid breaking behavior, I'd support both types, str and list (at least for a while). But this implies using the any type, or, move to JSON schema which allows multiple types AFAIK.

which json schema tool? worth a look at. But `any` type would work as well I guess with deprecation, and then we can switch to list again.

PerMac commented:
I agree that doing some cleaning and improving in .yamls would be beneficial. Another confusing property is that platform_allow and integration_platforms are of different types (str and sequence). 
Maybe we could get "multitypes" in a simplistic way? When we define the new schema modify twister to try-except check if changes are fulfilling new schema, and if exception is raised check for old schema. If old schema works -> Print msg that yaml is not compliant with the new schema but with the old, and warn that the old will be deprecated (or do nothing if we are fine with keeping both). 

When we were working on twister v2 we used marshmallow library to do the schemas and validation https://marshmallow.readthedocs.io/en/stable/why.html. We believe we got very close to what exists in the upstream. E.g. https://github.com/zephyrproject-rtos/twister/blob/main/src/twister2/yaml_test_specification.py From quick search it seems doing mulitype schema is not an issue with marshmallow https://stackoverflow.com/questions/61614546/python-marshmallow-field-can-be-two-different-types. 
But it is just another idea. 


To recap:
- I believe doing some "revolution" i.e. dropping the support for the existing yamls schemas would be quite problematic and cause some backlash. Therefore, both "legacy" and "new" will have to be supported at least for couple releases.
- Can the existing tools support multitype fields? If not, shall we look for sth else (e.g. json schema, marshmallow?) 
- I am against having platform_allow and allow_platforms as a mitigation for different types. It will introduce even more confusion

gmarull committed:
twister: set 'any' type for set/list fields

This allows to write both, lists or strings in the YAML file.
Unfortunately pykwalify schema language does not support multi-type
fields. JSON schema or Marshmallow could be used instead.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
twister: improve diagnostics for 'any' set/list types

Since pykwalify does not support multi-type fields, we are forced to use
the 'any' type. This is used for fields that allow sets/lists written as
either YAML lists or space-separated strings. This patch checks for
str/list Python type, and raises a configuration error if not true so
that user obtains better error messages in case the configuration file
contains invalid format.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
scripts: utils: add twister_to_list tool

Add a new tool that convers string-based lists on Twister config files
to native YAML lists.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: lib: mpsc_pbuf: remove duplicate timeout key

Test had 2 timeout entries (90, 120). Left 120 to be safe, as it's the
larger value.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples, tests: convert string-based twister lists to YAML lists

Twister now supports using YAML lists for all fields that were written
as space-separated lists. Used twister_to_list.py script. Some artifacts
on string length are due to how ruamel dumps content.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
twister: deprecate space-separated lists

Deprecate in favor of native YAML lists.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
twister: handle mixed common/tests types

This patch handles mixed types between common/tests sections in the
configuration, e.g. str+list, list+str, str+str.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
Updated PR:

- All space-separated list fields can now be specified as YAML lists. For single items, I think we should continue supporting strings, for simplicity.
- Provided a script to automatically migrate twister config files
- Migrated all in-tree files using the script
- Space-separated lists are deprecated (I used a Python deprecation warning, but I can't see it when running Twister, should likely be improved?)

cfriedt commented:
LGTM - just a couple of linter issues

de-nordic reviewed (dismissed):
MCUmgr tests, fs and flash tests look OK.

henrikbrixandersen reviewed (dismissed):
LGTM. Was the conversion done programmatically?

gmarull commented:
> LGTM - just a couple of linter issues. I guess there would need to be some kind of deprecation process for the older style.

For now I've flagged the support for string-based list as deprecated in Twister's config parser.

aescolar commented:
This PR is quite a merge conflict generator. I'd merge it soonish than later
